base/shell/explorer/trayntfy.cpp

```diff
@@ -33,6 +34,255 @@ typedef struct _SYS_PAGER_COPY_DATA
     NOTIFYICONDATA  nicon_data;
 } SYS_PAGER_COPY_DATA, *PSYS_PAGER_COPY_DATA;
 
+
+struct IconWatcherData
+{
+    HANDLE hProcess;
+    DWORD ProcessId;
+    NOTIFYICONDATA IconData;
+
+    IconWatcherData(NOTIFYICONDATA *iconData) :
+        hProcess(NULL), ProcessId(0)
+    {
+        IconData.cbSize = sizeof(NOTIFYICONDATA);
+        IconData.hWnd = iconData->hWnd;
+        IconData.uID = iconData->uID;
+        IconData.guidItem = iconData->guidItem;
+    }
+
+    ~IconWatcherData()
+    {
+        if (hProcess)
+        {
+            CloseHandle(hProcess);
+        }
+    }
+};
+
+class CIconWatcher
+{
+    CAtlList<IconWatcherData *> m_WatcherList;
```

Comment:
Why does this need to keep its own list of icons when the toolbar already has them all?

Reply:
Yeah I'd considered that, but its own list of items uses a public struct (NOTIFYICONDATA), and I needed to add a process handle and pid.
It was simpler to just keep everything self contained in this class than to start modifying existing code to store a NOTIFYICONDATA plus additional stuff. Also the existing code was reliant on the non-reentrancy of the message loop, so didn't use locking. My code was orignally accessing data outside of this and needed to be thread safe .... although I eventually changed the 'delete' routine to call SendMessage, so that became less of an issue.

Reply:
In my local copy, where I started implementing versioning differences, I had to find a way to save that version number (since uVersion is unioned with uTimeout, and we need to keep both), so I did this:

struct InternalIconData : NOTIFYICONDATA
{
    UINT uVersionCopy;
};

And then changed all the uses of the toolbar data, to InternalIconData

Reply:
See: https://github.com/gigaherz/reactos/commit/acec739a350087cfa5f1fddaadc0bcc13aa45b53

Reply:
yeah, that was an option I toyed around with. Two different ways of achieving the same result I suppose. I just felt this was was slightly cleaner in this situation as I was implementing a 'new' feature and wanted to keep it contained in a single class.

I can change it to modify existing data if people prefer?

Reply:
I don't mind either way. I was just mentioning we will most probably get this extended struct soonish, so it is not an issue at all do to that.